retention lease background sync failed after deleting index

<!--
GitHub is reserved for bug reports and feature requests; it is not the place
for general questions. If you have a question or an unconfirmed bug , please
visit the [forums](https://discuss.elastic.co/c/elasticsearch).  Please also
check your OS is [supported](https://www.elastic.co/support/matrix#show_os).
If it is not, the issue is likely to be closed.

For security vulnerabilities please only send reports to security@elastic.co.
See https://www.elastic.co/community/security for more information.

Please fill in the following details to help us reproduce the bug:
-->

**Elasticsearch version** (`bin/elasticsearch --version`): 7.7.0

**Plugins installed**: [`repository-s3`]

**JVM version** (`java -version`): bundled (i.e. 14)

**OS version** (`uname -a` if on a Unix-like system): `Linux/4.15.0-1058-aws/amd64`

**Description of the problem including expected versus actual behavior**:

When deleting an index which contains data (in our case it was 100GB large), we see the following exception trace:

```
[2020-06-09T07:27:37,710][INFO ][o.e.c.m.MetaDataDeleteIndexService] [elasticsearch-0] [elasticlogs-2020-01-01/5MAgXrFzTkaB3sILV7dqsA] deleting index
[2020-06-09T07:27:42,199][WARN ][o.e.i.s.RetentionLeaseSyncAction] [elasticsearch-0] [elasticlogs-2020-01-01][3] retention lease background sync failed
org.elasticsearch.transport.RemoteTransportException: [elasticsearch-0][10.10.1.56:9300][indices:admin/seq_no/retention_lease_background_sync[p]]
Caused by: org.elasticsearch.index.IndexNotFoundException: no such index [elasticlogs-2020-01-01]
        at org.elasticsearch.indices.IndicesService.indexServiceSafe(IndicesService.java:504) ~[elasticsearch-7.7.0.jar:7.7.0]
        at org.elasticsearch.action.support.replication.TransportReplicationAction.getIndexShard(TransportReplicationAction.java:591) ~[elasticsearch-7.7.0.jar:7.7.0]
[...]
```

While it is harmless that the retention lease background sync fails for an index that has just been deleted, the expectation is that the retention lease background sync is cancelled when the index is deleted (or it does not have any issues when the index is deleted).

**Steps to reproduce**:

1. Create an index with data in it (in our case it was 100GB large)
2. Delete the index
3. Observe the retention lease background sync failing in the logs

**Provide logs (if relevant)**:

[Full log file](https://github.com/elastic/elasticsearch/files/4750806/node.log) attached.

--- a/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncAction.java
+++ b/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncAction.java
@@ -39,6 +39,7 @@ import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.gateway.WriteStateException;
+import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.IndexShardClosedException;
 import org.elasticsearch.index.shard.ShardId;
@@ -137,8 +138,11 @@ public class RetentionLeaseBackgroundSyncAction extends TransportReplicationActi
                             // node shutting down
                             return;
                         }
-                        if (ExceptionsHelper.unwrap(e, AlreadyClosedException.class, IndexShardClosedException.class) != null) {
-                            // the shard is closed
+                        if (ExceptionsHelper.unwrap(e,
+                                                    IndexNotFoundException.class,
+                                                    AlreadyClosedException.class,
+                                                    IndexShardClosedException.class) != null) {
+                            // the index was deleted or the shard is closed
                             return;
                         }
                         getLogger().warn(new ParameterizedMessage("{} retention lease background sync failed", shardId), e);

--- a/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java
+++ b/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java
@@ -41,6 +41,7 @@ import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.gateway.WriteStateException;
+import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.IndexShardClosedException;
 import org.elasticsearch.index.shard.ShardId;
@@ -130,7 +131,10 @@ public class RetentionLeaseSyncAction extends
 
                     @Override
                     public void handleException(TransportException e) {
-                        if (ExceptionsHelper.unwrap(e, AlreadyClosedException.class, IndexShardClosedException.class) == null) {
+                        if (ExceptionsHelper.unwrap(e,
+                                                    IndexNotFoundException.class,
+                                                    AlreadyClosedException.class,
+                                                    IndexShardClosedException.class) == null) {
                             getLogger().warn(new ParameterizedMessage("{} retention lease sync failed", shardId), e);
                         }
                         task.setPhase("finished");

